juju/juju
# Release storage client worker provider

## Description of change

Add api, cmd, worker/storageprovisioner, and storage provider changes required to support non-destructively removing storage from a model.

The `juju remove-storage` command grows a `--no-destroy` flag, which instructs Juju to remove the storage without destroying the associated cloud storage resources. This can be used to release the storage from Juju's control. Once the storage is removed from the model, destroying the model or controller will not destroy the storage.

The VolumeSource and FilesystemSource interfaces grow methods for releasing storage, rather than destroying it. The storageprovisioner worker uses these methods for storage that is marked for releasing.

Requires https://github.com/juju/juju/pull/7648

## QA steps

1. juju bootstrap localhost
2. juju deploy postgresql --storage pgdata=lxd
3. juju remove-storage --force --no-destroy pgdata/0
4. juju destroy-controller localhost
5. lxc storage volume show juju juju-<model-uuid-suffix>-filesystem-0 # use "list" to get the name
(volume should remain, but should show no juju-controller-uuid or juju-model-uuid tags)

As above, but with EBS. In the portal, you should see the juju-controller-uuid/juju-model-uuid tags are empty after the storage is released/removed.

## Documentation changes

Yes, we'll need to document `juju remove-storage --no-destroy`.

## Bug reference

None.
